File: go_wails_react/backend/db_manager.go
// Gerenciador de conexoes DuckDB em Go - carrega bancos, lista tabelas, executa queries
// !T: Core database manager handling DuckDB connections, query execution and schema inspection

package backend

import (
	"database/sql"
	"fmt"

	_ "github.com/marcboeker/go-duckdb" // !T: Import for side effects (register DuckDB driver)
)

// Struct que gerencia acesso aos bancos DuckDB - conectar, listar tabelas, queries
// !T: Database manager with singleton pattern holding connection and current DB path
type DBManager struct {
	conn      *sql.DB // !T: Standard sql.DB interface from database/sql
	currentDB string
	tableSet  map[string]struct{}
	tableList []string
}

// Cria novo DBManager sem conectar - conexao sera feita com Connect()
// !T: Factory function returning unconnected DBManager, lazy initialization pattern
func NewDBManager() *DBManager {
	return &DBManager{}
}

// Conecta ao banco DuckDB especificado - fecha conexao anterior se existir
// !T: Opens read-only connection with DSN format: path?access_mode=read_only
func (m *DBManager) Connect(dbPath string) error {
	// Fecha conexao anterior se existir
	if m.conn != nil {
		m.conn.Close()
	}

	// Abre conexao DuckDB com modo read-only
	// !T: DSN format: path?access_mode=read_only prevents accidental writes
	dsn := fmt.Sprintf("%s?access_mode=read_only", dbPath)
	conn, err := sql.Open("duckdb", dsn)
	if err != nil {
		return fmt.Errorf("failed to open database: %w", err)
	}

	// Testa se conexao funciona
	if err := conn.Ping(); err != nil {
		conn.Close()
		return fmt.Errorf("failed to ping database: %w", err)
	}

	m.conn = conn
	m.currentDB = dbPath

	// Pre-carrega cache de tabelas para validar nomes sem overhead por chamada
	if err := m.refreshTableCache(); err != nil {
		// fecha conexao recem aberta para evitar estado inconsistente
		conn.Close()
		m.conn = nil
		m.currentDB = ""
		return fmt.Errorf("failed to load table cache: %w", err)
	}
	return nil
}

// Retorna lista de nomes de tabelas no banco atual
// !T: Executes DuckDB-specific SHOW TABLES query and returns string slice
func (m *DBManager) ListTables() ([]string, error) {
	if m.conn == nil {
		return nil, fmt.Errorf("no database connected")
	}

	// Usa cache preenchido no Connect; evita nova consulta se cache existir
	if len(m.tableList) > 0 && len(m.tableSet) == len(m.tableList) {
		copied := make([]string, len(m.tableList))
		copy(copied, m.tableList)
		return copied, nil
	}

	if err := m.refreshTableCache(); err != nil {
		return nil, err
	}
	copied := make([]string, len(m.tableList))
	copy(copied, m.tableList)
	return copied, nil
}

// Struct que representa schema de uma coluna com nome, tipo e nullable
// !T: Column metadata struct exported to frontend via JSON tags
type ColumnSchema struct {
	Name string `json:"name"`
	Type string `json:"type"`
	Null string `json:"null"`
}

// Retorna schema de uma tabela (colunas, tipos, nullable)
// !T: Executes PRAGMA table_info() which is more efficient than DESCRIBE SELECT *
func (m *DBManager) GetTableSchema(tableName string) ([]ColumnSchema, error) {
	if m.conn == nil {
		return nil, fmt.Errorf("no database connected")
	}

	// Valida tableName contra lista de tabelas
	// !T: Prevent SQL injection by validating input
	validatedName, err := m.validateTableName(tableName)
	if err != nil {
		return nil, err
	}
	tableName = validatedName

	// PRAGMA table_info retorna metadata sem varrer dados
	// !T: More efficient than DESCRIBE SELECT *
	query := fmt.Sprintf(`PRAGMA table_info("%s")`, tableName)
	rows, err := m.conn.Query(query)
	if err != nil {
		return nil, fmt.Errorf("failed to get table info: %w", err)
	}
	defer rows.Close()

	// Le metadata de cada coluna
	// !T: PRAGMA returns: cid, name, type, notnull, dflt_value, pk
	var schema []ColumnSchema
	for rows.Next() {
		var cid int
		var name, colType string
		var notNull int
		var dfltValue interface{}
		var pk int

		if err := rows.Scan(&cid, &name, &colType, &notNull, &dfltValue, &pk); err != nil {
			return nil, fmt.Errorf("failed to scan column: %w", err)
		}

		nullStr := "YES"
		if notNull == 1 {
			nullStr = "NO"
		}

		schema = append(schema, ColumnSchema{
			Name: name,
			Type: colType,
			Null: nullStr,
		})
	}

	return schema, rows.Err()
}

// Valida se tableName existe no banco
// !T: Prevents SQL injection using parameterized query against information_schema
func (m *DBManager) validateTableName(tableName string) (string, error) {
	if m.conn == nil {
		return "", fmt.Errorf("no database connected")
	}

	// Usa cache para evitar overhead de conexao repetida
	if m.tableSet == nil || len(m.tableSet) == 0 {
		if err := m.refreshTableCache(); err != nil {
			return "", err
		}
	}
	if _, ok := m.tableSet[tableName]; !ok {
		return "", fmt.Errorf("table not found: %s", tableName)
	}

	return tableName, nil
}

// Executa SELECT na tabela e retorna resultados como slice de maps
// !T: Returns []map[string]interface{} for dynamic column handling in frontend
func (m *DBManager) QueryTable(tableName string, limit int) ([]map[string]interface{}, error) {
	if m.conn == nil {
		return nil, fmt.Errorf("no database connected")
	}

	// Valida tableName contra lista de tabelas
	// !T: Prevent SQL injection by validating input
	validatedName, err := m.validateTableName(tableName)
	if err != nil {
		return nil, err
	}
	tableName = validatedName

	// SELECT com LIMIT parametrizado
	// !T: Only limit is parameterized, table name is validated
	query := fmt.Sprintf(`SELECT * FROM "%s" LIMIT ?`, tableName)
	rows, err := m.conn.Query(query, limit)
	if err != nil {
		return nil, fmt.Errorf("failed to query table: %w", err)
	}
	defer rows.Close()

	// Pega nomes das colunas
	columns, err := rows.Columns()
	if err != nil {
		return nil, fmt.Errorf("failed to get columns: %w", err)
	}

	// Le cada row e converte para map
	var results []map[string]interface{}
	for rows.Next() {
		// !T: Create slice of interface{} to scan into
		values := make([]interface{}, len(columns))
		valuePtrs := make([]interface{}, len(columns))
		for i := range values {
			valuePtrs[i] = &values[i]
		}

		if err := rows.Scan(valuePtrs...); err != nil {
			return nil, fmt.Errorf("failed to scan row: %w", err)
		}

		// Cria map com nome_coluna -> valor
		row := make(map[string]interface{})
		for i, col := range columns {
			row[col] = values[i]
		}
		results = append(results, row)
	}

	return results, rows.Err()
}

// Retorna numero total de linhas em uma tabela
// !T: Executes COUNT(*) query with validated table name
func (m *DBManager) GetRowCount(tableName string) (int, error) {
	if m.conn == nil {
		return 0, fmt.Errorf("no database connected")
	}

	// Valida tableName contra lista de tabelas
	// !T: Prevent SQL injection by validating input
	validatedName, err := m.validateTableName(tableName)
	if err != nil {
		return 0, err
	}
	tableName = validatedName

	// COUNT(*) retorna total de linhas
	// !T: Table name validated, safe to use in query
	query := fmt.Sprintf(`SELECT COUNT(*) FROM "%s"`, tableName)
	var count int
	if err = m.conn.QueryRow(query).Scan(&count); err != nil {
		return 0, fmt.Errorf("failed to count rows: %w", err)
	}

	return count, nil
}

// Fecha conexao com o banco - sempre chame ao finalizar uso
// !T: Closes sql.DB connection and resets manager state
func (m *DBManager) Close() error {
	if m.conn != nil {
		err := m.conn.Close()
		m.conn = nil
		m.currentDB = ""
		m.tableSet = nil
		m.tableList = nil
		return err
	}
	return nil
}

// Atualiza cache de tabelas existentes no banco atual
func (m *DBManager) refreshTableCache() error {
	if m.conn == nil {
		return fmt.Errorf("no database connected")
	}

	rows, err := m.conn.Query("SHOW TABLES")
	if err != nil {
		return fmt.Errorf("failed to list tables: %w", err)
	}
	defer rows.Close()

	tableSet := make(map[string]struct{})
	var tableList []string
	for rows.Next() {
		var tableName string
		if err := rows.Scan(&tableName); err != nil {
			return fmt.Errorf("failed to scan table name: %w", err)
		}
		tableSet[tableName] = struct{}{}
		tableList = append(tableList, tableName)
	}
	if err := rows.Err(); err != nil {
		return err
	}

	m.tableSet = tableSet
	m.tableList = tableList
	return nil
}

File: py_qt6/src/backend/db_manager.py
# Gerenciador de conexoes DuckDB - carrega bancos, lista tabelas, executa queries
# !T: DuckDB connection manager with read-only mode and metadata caching

import duckdb
from pathlib import Path
from typing import Optional, List, Dict, Any

# !T: DuckDB connection manager with read-only mode
# !T: Handles multiple database attachments and metadata caching


class DBManager:
    """
    Classe que gerencia acesso aos bancos DuckDB convertidos.

    Funcionalidades:
    - Conectar ao banco principal (ultimo convertido)
    - Listar tabelas disponiveis
    - Executar queries e retornar resultados como lista de dicts
    - Fechar conexao de forma segura
    """

    def __init__(self, db_path: Optional[str] = None):
        # Inicializa manager sem conectar - db_path pode ser None
        self.conn: Optional[duckdb.DuckDBPyConnection] = None
        self.current_db: Optional[str] = None

        if db_path:
            self.connect(db_path)

    def connect(self, db_path: str) -> None:
        """
        Conecta ao banco DuckDB especificado - fecha conexao anterior se existir.

        Args:
            db_path: Caminho para arquivo .duckdb
        """
        # !T: Close existing connection to avoid resource leak
        if self.conn:
            self.conn.close()

        # Valida caminho do banco
        # !T: Prevent path traversal and validate file extension
        validated_path = self._validate_database_path(db_path)

        # Abre conexao read-only para nao modificar dados
        # !T: read_only=True prevents accidental writes
        self.conn = duckdb.connect(validated_path, read_only=True)
        self.current_db = validated_path

    def _validate_database_path(self, db_path: str) -> str:
        """
        Valida caminho do banco de dados contra path traversal.
        !T: Prevents path traversal and validates file extension.

        Args:
            db_path: Caminho para validar

        Returns:
            Caminho absoluto validado

        Raises:
            ValueError: Se caminho invalido
            FileNotFoundError: Se arquivo nao existe
        """
        # Converte para caminho absoluto
        abs_path = Path(db_path).resolve()

        # Valida extensao do arquivo
        # !T: Only .duckdb and .db files allowed
        if abs_path.suffix not in [".duckdb", ".db"]:
            raise ValueError(f"Only .duckdb and .db files are supported, got: {abs_path.suffix}")

        # Verifica se arquivo existe
        if not abs_path.exists():
            raise FileNotFoundError(f"Database not found: {abs_path}")

        return str(abs_path)

    def list_tables(self) -> List[str]:
        """
        Retorna lista de nomes de tabelas no banco atual.

        Returns:
            Lista de strings com nomes das tabelas
        """
        if not self.conn:
            return []

        # Query SHOW TABLES retorna DataFrame
        # !T: DuckDB returns pandas DataFrame by default
        result = self.conn.execute("SHOW TABLES").df()

        # Converte coluna 'name' para lista Python
        return result["name"].tolist()

    def _validate_table_name(self, table_name: str) -> None:
        """
        Valida se tableName existe no banco.
        !T: Prevents SQL injection by checking against known tables.

        Args:
            table_name: Nome da tabela para validar

        Raises:
            ValueError: Se tabela nao existir
        """
        tables = self.list_tables()
        if table_name not in tables:
            raise ValueError(f"Table not found: {table_name}")

    def get_table_schema(self, table_name: str) -> List[Dict[str, Any]]:
        """
        Retorna schema de uma tabela (colunas, tipos, nullable).

        Args:
            table_name: Nome da tabela

        Returns:
            Lista de dicts com column_name, column_type, null
        """
        if not self.conn:
            return []

        # Valida tableName contra lista de tabelas
        # !T: Prevent SQL injection by validating input
        self._validate_table_name(table_name)

        # PRAGMA table_info retorna metadata sem varrer dados
        # !T: More efficient than DESCRIBE SELECT *
        result = self.conn.execute(
            f'PRAGMA table_info("{table_name}")'
        ).fetchall()

        # Converte tuplas em lista de dicts
        # !T: PRAGMA returns: cid, name, type, notnull, dflt_value, pk
        return [
            {
                "column_name": row[1],  # name
                "column_type": row[2],  # type
                "null": "NO" if row[3] == 1 else "YES",  # notnull
            }
            for row in result
        ]

    def query_table(
        self, table_name: str, limit: int = 100
    ) -> List[Dict[str, Any]]:
        """
        Executa SELECT na tabela e retorna resultados.

        Args:
            table_name: Nome da tabela
            limit: Numero maximo de linhas (padrao 100)

        Returns:
            Lista de dicts, cada dict = uma linha
        """
        if not self.conn:
            return []

        # Valida tableName contra lista de tabelas
        # !T: Prevent SQL injection by validating input
        self._validate_table_name(table_name)

        # SELECT com LIMIT para nao carregar tabela inteira
        # !T: Table name validated, safe to use in query
        query = f'SELECT * FROM "{table_name}" LIMIT {limit}'
        result = self.conn.execute(query).df()

        # Converte DataFrame pandas para lista de dicts
        # !T: to_dict('records') returns list of row dicts
        return result.to_dict("records")

    def execute_query(self, query: str) -> List[Dict[str, Any]]:
        """
        Executa query SQL customizada.

        Args:
            query: Query SQL completa

        Returns:
            Lista de dicts com resultados
        """
        if not self.conn:
            return []

        # !T: Execute arbitrary SQL and return as records
        result = self.conn.execute(query).df()
        return result.to_dict("records")

    def get_row_count(self, table_name: str) -> int:
        """
        Retorna numero total de linhas em uma tabela.

        Args:
            table_name: Nome da tabela

        Returns:
            Numero de linhas (int)
        """
        if not self.conn:
            return 0

        # Valida tableName contra lista de tabelas
        # !T: Prevent SQL injection by validating input
        self._validate_table_name(table_name)

        # COUNT(*) retorna total de linhas
        # !T: Table name validated, safe to use in query
        query = f'SELECT COUNT(*) as count FROM "{table_name}"'
        result = self.conn.execute(query).fetchone()

        return result[0] if result else 0

    def close(self) -> None:
        """
        Fecha conexao com o banco - sempre chame ao finalizar uso.
        """
        if self.conn:
            # !T: Explicitly close connection to release file handle
            self.conn.close()
            self.conn = None
            self.current_db = None

    def __enter__(self):
        # Suporte a context manager (with statement)
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        # Fecha conexao automaticamente ao sair do with
        # !T: Context manager is the reliable way to manage resources
        self.close()
